<a href="https://colab.research.google.com/github/YuriSerhiienko/data_science_hw10/blob/main/hw10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import tensorflow as tf
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from tensorflow.keras.models import Sequential
from tensorflow.keras.datasets import fashion_mnist
from sklearn.metrics import classification_report, confusion_matrix
import numpy as np

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Нормалізація зображень
train_images = train_images / 255.0
test_images = test_images / 255.0

# Зміна розмірів зображень для CNN
train_images = train_images.reshape((train_images.shape[0], 28, 28, 1))
test_images = test_images.reshape((test_images.shape[0], 28, 28, 1))

model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10, validation_split=0.2)

test_loss, test_acc = model.evaluate(test_images, test_labels)
print('\nТочність на тестових даних:', test_acc)

predictions = model.predict(test_images)
y_pred = np.argmax(predictions, axis=1)

print("\nМатриця помилок:")
print(confusion_matrix(test_labels, y_pred))

print("\nКласифікаційний звіт:")
print(classification_report(test_labels, y_pred))


Epoch 1/10
1500/1500 [==============================] - 8s 4ms/step - loss: 0.4683 - accuracy: 0.8298 - val_loss: 0.3533 - val_accuracy: 0.8733
Epoch 2/10
1500/1500 [==============================] - 5s 4ms/step - loss: 0.3108 - accuracy: 0.8876 - val_loss: 0.2913 - val_accuracy: 0.8939
Epoch 3/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.2645 - accuracy: 0.9035 - val_loss: 0.2891 - val_accuracy: 0.8908
Epoch 4/10
1500/1500 [==============================] - 5s 4ms/step - loss: 0.2323 - accuracy: 0.9142 - val_loss: 0.2651 - val_accuracy: 0.9055
Epoch 5/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.2048 - accuracy: 0.9244 - val_loss: 0.2686 - val_accuracy: 0.9046
Epoch 6/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.1802 - accuracy: 0.9327 - val_loss: 0.2445 - val_accuracy: 0.9116
Epoch 7/10
1500/1500 [==============================] - 5s 4ms/step - loss: 0.1593 - accuracy: 0.9403 - val_loss: 0.2660 - val_accuracy:

In [1]:
import tensorflow as tf
from tensorflow.keras.applications import VGG16
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Dropout, Flatten
from tensorflow.keras.datasets import fashion_mnist
from sklearn.metrics import classification_report, confusion_matrix
import numpy as np

# Завантаження та підготовка датасету
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
train_images = train_images / 255.0
test_images = test_images / 255.0

# Зміна розмірів зображень для VGG16
def resize_images(images):
    images_resized = np.array([tf.image.resize(img, (32, 32)) for img in images])
    return images_resized

train_images_resized = resize_images(train_images[..., np.newaxis])
test_images_resized = resize_images(test_images[..., np.newaxis])

# Конвертація зображень у формат RGB
train_images_rgb = np.repeat(train_images_resized, 3, axis=-1)
test_images_rgb = np.repeat(test_images_resized, 3, axis=-1)

# Завантаження VGG16
vgg_base = VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))
vgg_base.trainable = False  # Заморожування ваг

# Створення моделі
model = tf.keras.Sequential([
    vgg_base,
    Flatten(),
    Dense(128, activation='relu'),
    Dropout(0.5),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Навчання моделі
history = model.fit(train_images_rgb, train_labels, epochs=10, validation_split=0.2)

# Оцінка моделі
test_loss, test_acc = model.evaluate(test_images_rgb, test_labels)
print('\nТочність на тестових даних:', test_acc)

# Передбачення та метрики якості
predictions = model.predict(test_images_rgb)
y_pred = np.argmax(predictions, axis=1)

print("\nМатриця помилок:")
print(confusion_matrix(test_labels, y_pred))

print("\nКласифікаційний звіт:")
print(classification_report(test_labels, y_pred))


58889256/58889256 [==============================] - 0s 0us/step
Epoch 1/10
1500/1500 [==============================] - 19s 10ms/step - loss: 0.6978 - accuracy: 0.7533 - val_loss: 0.4747 - val_accuracy: 0.8224
Epoch 2/10
1500/1500 [==============================] - 14s 10ms/step - loss: 0.5068 - accuracy: 0.8180 - val_loss: 0.4394 - val_accuracy: 0.8384
Epoch 3/10
1500/1500 [==============================] - 18s 12ms/step - loss: 0.4708 - accuracy: 0.8303 - val_loss: 0.4269 - val_accuracy: 0.8422
Epoch 4/10
1500/1500 [==============================] - 15s 10ms/step - loss: 0.4462 - accuracy: 0.8393 - val_loss: 0.4096 - val_accuracy: 0.8477
Epoch 5/10
1500/1500 [==============================] - 17s 11ms/step - loss: 0.4281 - accuracy: 0.8461 - val_loss: 0.4006 - val_accuracy: 0.8492
Epoch 6/10
1500/1500 [==============================] - 15s 10ms/step - loss: 0.4185 - accuracy: 0.8479 - val_loss: 0.3954 - val_accuracy: 0.8539
Epoch 7/10
1500/1500 [==============================] - 17s

Попередня робота (Багатошаровий перцептрон)

Архітектура: Базувалася на повнозв'язних шарах (Dense layers).

Точність на тестових даних: 88.08% у базовій моделі, 86.93% після оновлення архітектури.

Інші метрики: Висока точність (precision) та повнота (recall), особливо після оновлення архітектури.

Нинішня робота (Згорткова нейронна мережа, CNN)

Архітектура: Базувалася на згорткових шарах (Conv2D), що є типовим для задач обробки зображень.

Точність на тестових даних: 90.88%.

Інші метрики: Деталізовані метрики (матриця помилок, точність, повнота, F1-оцінка) показали високу ефективність моделі у класифікації зображень.

Висновки

Ефективність архітектури: Згорткова нейронна мережа (CNN) показала кращі результати для задачі класифікації зображень у порівнянні з багатошаровим перцептроном. Це очікувано, оскільки CNN краще адаптовані для виявлення шаблонів у зображеннях.

Підхід до обробки даних: CNN використовує просторові відносини між пікселями, що дозволяє ефективно виділяти особливості зображень, в той час як багатошаровий перцептрон обробляє дані як одновимірний масив.



In [2]:
import tensorflow as tf
from tensorflow.keras.applications import VGG16
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Dropout, Flatten
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.utils import to_categorical
from sklearn.metrics import classification_report, confusion_matrix
import numpy as np

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Нормалізація зображень
train_images = train_images / 255.0
test_images = test_images / 255.0

# Перетворення міток у one-hot вектори
train_labels = to_categorical(train_labels, 10)
test_labels = to_categorical(test_labels, 10)

train_images = np.array([np.resize(img, (32, 32, 3)) for img in train_images])
test_images = np.array([np.resize(img, (32, 32, 3)) for img in test_images])

vgg_base = VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))
vgg_base.trainable = False  # Заморожування ваг VGG16

model = tf.keras.Sequential([
    vgg_base,
    Flatten(),
    Dense(128, activation='relu'),
    Dropout(0.5),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(train_images, train_labels, epochs=10, validation_split=0.2)

test_loss, test_acc = model.evaluate(test_images, test_labels)
print('\nТочність на тестових даних:', test_acc)

predictions = model.predict(test_images)
y_pred = np.argmax(predictions, axis=1)
test_labels_argmax = np.argmax(test_labels, axis=1)

print("\nМатриця помилок:")
print(confusion_matrix(test_labels_argmax, y_pred))

print("\nКласифікаційний звіт:")
print(classification_report(test_labels_argmax, y_pred))


Epoch 1/10
1500/1500 [==============================] - 19s 12ms/step - loss: 1.1109 - accuracy: 0.6114 - val_loss: 0.8442 - val_accuracy: 0.7138
Epoch 2/10
1500/1500 [==============================] - 17s 12ms/step - loss: 0.8772 - accuracy: 0.6989 - val_loss: 0.7665 - val_accuracy: 0.7370
Epoch 3/10
1500/1500 [==============================] - 15s 10ms/step - loss: 0.8186 - accuracy: 0.7187 - val_loss: 0.7305 - val_accuracy: 0.7452
Epoch 4/10
1500/1500 [==============================] - 15s 10ms/step - loss: 0.7913 - accuracy: 0.7276 - val_loss: 0.7098 - val_accuracy: 0.7533
Epoch 5/10
1500/1500 [==============================] - 15s 10ms/step - loss: 0.7688 - accuracy: 0.7344 - val_loss: 0.7042 - val_accuracy: 0.7557
Epoch 6/10
1500/1500 [==============================] - 16s 10ms/step - loss: 0.7518 - accuracy: 0.7401 - val_loss: 0.6769 - val_accuracy: 0.7605
Epoch 7/10
1500/1500 [==============================] - 15s 10ms/step - loss: 0.7387 - accuracy: 0.7415 - val_loss: 0.6859 -

In [6]:
import tensorflow as tf
import numpy as np
from tensorflow.keras.applications import VGG16
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Flatten, GlobalAveragePooling2D
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.optimizers import Adam
import cv2

# Завантаження та підготовка датасету
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Зміна розміру зображень до 32x32 та конвертація у формат RGB
train_images_rgb = np.array([cv2.cvtColor(cv2.resize(img, (32, 32)), cv2.COLOR_GRAY2RGB) for img in train_images]) / 255.0
test_images_rgb = np.array([cv2.cvtColor(cv2.resize(img, (32, 32)), cv2.COLOR_GRAY2RGB) for img in test_images]) / 255.0

# Перетворення міток у one-hot вектори
train_labels = to_categorical(train_labels, 10)
test_labels = to_categorical(test_labels, 10)

# Завантаження VGG16
vgg_base = VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))

# Розморожування останніх шарів VGG16
for layer in vgg_base.layers[:-4]:
    layer.trainable = False

# Створення кастомізованої моделі
model = Sequential([
    vgg_base,
    GlobalAveragePooling2D(),
    Dense(1024, activation='relu'),
    Dropout(0.5),
    Dense(512, activation='relu'),
    Dropout(0.5),
    Dense(10, activation='softmax')
])

# Компіляція моделі
model.compile(optimizer=Adam(learning_rate=0.0001),
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Навчання моделі
history = model.fit(train_images_rgb, train_labels, epochs=20, batch_size=32, validation_split=0.2)

# Оцінка моделі
test_loss, test_acc = model.evaluate(test_images_rgb, test_labels)
print('\nТочність на тестових даних:', test_acc)

# Передбачення та отримання міток класів
predictions = model.predict(test_images_rgb)
predicted_classes = np.argmax(predictions, axis=1)
true_classes = np.argmax(test_labels, axis=1)

# Матриця помилок
from sklearn.metrics import confusion_matrix, classification_report
print("\nМатриця помилок:")
print(confusion_matrix(true_classes, predicted_classes))

# Класифікаційний звіт
print("\nКласифікаційний звіт:")
print(classification_report(true_classes, predicted_classes))


Epoch 1/20
1500/1500 [==============================] - 29s 18ms/step - loss: 0.4962 - accuracy: 0.8255 - val_loss: 0.3332 - val_accuracy: 0.8804
Epoch 2/20
1500/1500 [==============================] - 29s 19ms/step - loss: 0.3209 - accuracy: 0.8871 - val_loss: 0.3156 - val_accuracy: 0.8832
Epoch 3/20
1500/1500 [==============================] - 27s 18ms/step - loss: 0.2804 - accuracy: 0.8999 - val_loss: 0.2893 - val_accuracy: 0.8957
Epoch 4/20
1500/1500 [==============================] - 26s 18ms/step - loss: 0.2557 - accuracy: 0.9079 - val_loss: 0.2617 - val_accuracy: 0.9035
Epoch 5/20
1500/1500 [==============================] - 26s 18ms/step - loss: 0.2312 - accuracy: 0.9176 - val_loss: 0.2617 - val_accuracy: 0.9068
Epoch 6/20
1500/1500 [==============================] - 26s 18ms/step - loss: 0.2132 - accuracy: 0.9222 - val_loss: 0.2685 - val_accuracy: 0.9053
Epoch 7/20
1500/1500 [==============================] - 29s 19ms/step - loss: 0.1985 - accuracy: 0.9284 - val_loss: 0.2738 -

CNN показала значно кращі результати в порівнянні з багатошаровою мережею. Точність CNN була вищою і становила близько 91%, в той час як точність багатошарової мережі була нижчою і становила близько 87%. Це означає, що згорткова мережа ефективніше впоралася з класифікацією зображень Fashion MNIST.

Висновок полягає в тому, що для обробки зображень, особливо для завдань класифікації зображень, CNN може бути кращим вибором порівняно з багатошаровими мережами. Використання згорткових шарів дозволяє моделі автоматично виявляти важливі ознаки на зображеннях і покращує її результати.